cmd/helm/chart_login.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/pkg/term"
+	"github.com/spf13/cobra"
+
+	"helm.sh/helm/cmd/helm/require"
+	"helm.sh/helm/pkg/action"
+)
+
+const chartLoginDesc = `
+Authenticate to a remote registry.
+`
+
+func newChartLoginCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
+	var usernameOpt, passwordOpt string
+	var passwordFromStdinOpt bool
+
+	cmd := &cobra.Command{
+		Use:   "login [host]",
+		Short: "login to a registry",
+		Long:  chartLoginDesc,
+		Args:  require.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			hostname := args[0]
+
+			username, password, err := getUsernamePassword(usernameOpt, passwordOpt, passwordFromStdinOpt)
+			if err != nil {
+				return err
+			}
+
+			return action.NewChartLogin(cfg).Run(out, hostname, username, password)
+		},
+	}
+
+	f := cmd.Flags()
+	f.StringVarP(&usernameOpt, "username", "u", "", "registry username")
+	f.StringVarP(&passwordOpt, "password", "p", "", "registry password or identity token")
+	f.BoolVarP(&passwordFromStdinOpt, "password-stdin", "", false, "read password or identity token from stdin")
+
+	return cmd
+}
+
+// Adapted from https://github.com/deislabs/oras
+func getUsernamePassword(usernameOpt string, passwordOpt string, passwordFromStdinOpt bool) (string, string, error) {
```

Comment:
just one usability suggestion: Docker cli recently had a change that if user tries to do docker login myregistry.io without user name and password, it first tries to read from the credential store and use the stored one for login. This comes very handy. Will helm login support the same?

Reply:
im not sure i understand - what is the "stored one", and why log into something that youve already authed against?

Reply:
The first time you do log in, the credential is stored; the next time you do login without username/pwd, the stored credential is retrieved and used for login.

The login in certain systems has certain significance. For example, when you do login, a token may be retrieved which is used in subsequent requests. But the token may expire after some time. Doing login again basically retrieve a new token.

Or server can do some special logic on login, say refreshing cached permission, etc.

Or I simply forgot if I did login already:)